Error while checking build/O2/fullCI for 658dc9472f4540219b92e0035e38ba038a501db2 at 2024-04-18 17:35:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/13038-slc8_x86-64/0/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx:87:25: error: no matching function for call to 'atan2(ROOT::Math::PositionVector3D<ROOT::Math::Cartesian3D<float>, ROOT::Math::DefaultCoordinateSystemTag>::Scalar)'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13038/658dc9472f4540219b92e0035e38ba038a501db2/build_O2_fullCI/pretty.html).

--- a/Detectors/MUON/MCH/Simulation/include/MCHSimulation/Hit.h
+++ b/Detectors/MUON/MCH/Simulation/include/MCHSimulation/Hit.h
@@ -25,19 +25,24 @@ class Hit : public ::o2::BasicXYZEHit<float>
 
  public:
   Hit(int trackId = 0, short detElemId = 0, math_utils::Point3D<float> entrancePoint = {}, const math_utils::Point3D<float> exitPoint = {},
-      float eloss = 0.0, float length = 0.0, float tof = 0.0)
-    : ::o2::BasicXYZEHit<float>(entrancePoint.x(), entrancePoint.y(), entrancePoint.z(), tof, eloss, trackId, detElemId), mLength{length}, mExitPoint(exitPoint)
+      float eloss = 0.0, float length = 0.0, float entranceTof = 0.0, float exitTof = 0.0)
+    : ::o2::BasicXYZEHit<float>(entrancePoint.x(), entrancePoint.y(), entrancePoint.z(), entranceTof, eloss, trackId, detElemId), mLength{length}, mExitPoint(exitPoint), mExitTof{exitTof}
   {
   }
 
   math_utils::Point3D<float> entrancePoint() const { return GetPos(); }
   math_utils::Point3D<float> exitPoint() const { return mExitPoint; }
 
+  //time in ns 
+  float entranceTof() const { return GetTime(); }
+  float exitTof() const { return mExitTof; }
+
   short detElemId() const { return GetDetectorID(); }
 
  private:
   float mLength = {};
   math_utils::Point3D<float> mExitPoint = {};
+  float mExitTof = {};
   ClassDefNV(Hit, 1);
 };
 

--- a/Detectors/MUON/MCH/Simulation/include/MCHSimulation/Response.h
+++ b/Detectors/MUON/MCH/Simulation/include/MCHSimulation/Response.h
@@ -32,7 +32,6 @@ class Response
  public:
   Response(Station station);
   ~Response() = default;
-
   float getChargeSpread() const { return mChargeSpread; }
   float getPitch() const { return mPitch; }
   float getSigmaIntegration() const { return mSigmaIntegration; }
@@ -48,6 +47,7 @@ class Response
    *
    */
   float etocharge(float edepos) const;
+
 
   /** Compute the charge fraction in a rectangle area for a unit charge
    * occuring at position (0,0)

--- a/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
@@ -104,7 +104,10 @@ void DEDigitizer::processHit(const Hit& hit, const InteractionRecord& collisionT
    //  field->Field(entrance, b);
   }
   //calculate track betagamma
-  auto betagamma = 50;//todo: use time of flight of entrance and exit
+  auto deltat = hit.exitTof() - hit.entranceTof();// delta t in nanoseconds
+  auto beta = std::sqrt((lexit.X() - lentrance.X()) * (lexit.X() - lentrance.X()) + (lexit.Y() - lentrance.Y()) * (lexit.Y() - lentrance.Y()) + hitlengthZ * hitlengthZ) / (deltat * 0.299792);
+
+  auto betagamma = beta * 1.0 / std::sqrt(1 - beta * beta);
   auto yAngleEffect = mResponse.inclandbfield(thetawire, betagamma, b[0]);
   localY += yAngleEffect;
 

--- a/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
@@ -17,6 +17,8 @@
 
 #include "DetectorsRaw/HBFUtils.h"
 #include "MCHSimulation/DigitizerParam.h"
+#include <TGeoGlobalMagField.h>
+#include "Field/MagneticField.h"
 
 #include "TRandom.h"
 
@@ -28,9 +30,11 @@ std::pair<o2::InteractionRecord, uint8_t> time2ROFtime(const o2::InteractionReco
   return std::make_pair(o2::InteractionRecord(time.bc - bc, time.orbit), bc);
 }
 
+//_________________________________________________________________________________________________
+
 namespace o2::mch
 {
-
+//_________________________________________________________________________________________________
 DEDigitizer::DEDigitizer(int deId, math_utils::Transform3D transformation, std::mt19937& random)
   : mDeId{deId},
     mResponse{deId < 300 ? Station::Type1 : Station::Type2345},
@@ -86,11 +90,21 @@ void DEDigitizer::processHit(const Hit& hit, const InteractionRecord& collisionT
   auto thetawire = asin((lexit.Y() - lentrance.Y()) / hitlengthZ);//check sign convention between O2 and Aliroot
 
   // bfield
-  float b[3] = {0., 0., 0.};
-  //  TGeoGlobalMagField::Instance()->Field(lentrance, b);//need to add Field as member as transformation or find other example (e.g. use TrackFitter.cxx as model)
-
+  double b[3] = {0., 0., 0.};
+  double x[3] = {lentrance.X(), lentrance.Y(), lentrance.Z()};
+  if (TGeoGlobalMagField::Instance()->GetField())  {
+    TGeoGlobalMagField::Instance()->Field(x, b);
+  } else{
+    LOG(fatal) << "no b field in DEDigitizer";
+   // auto l3Current = ic.options().get<float>("l3Current");
+   // auto dipoleCurrent = ic.options().get<float>("dipoleCurrent");
+   // auto field =
+   // o2::field::MagneticField::createFieldMap(l3Current, dipoleCurrent, o2::field::MagneticField::kConvLHC, false, 3500.,
+   //                                          "A-A", "$(O2_ROOT)/share/Common/maps/mfchebKGI_sym.root");
+   //  field->Field(entrance, b);
+  }
   //calculate track betagamma
-  auto betagamma = 1.;//todo: access track information
+  auto betagamma = 50;//todo: use time of flight of entrance and exit
   auto yAngleEffect = mResponse.inclandbfield(thetawire, betagamma, b[0]);
   localY += yAngleEffect;
 

--- a/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
@@ -86,26 +86,20 @@ void DEDigitizer::processHit(const Hit& hit, const InteractionRecord& collisionT
   auto localX = mResponse.getAnod(lpos.X());
   auto localY = lpos.Y();
 
-  //calculate angle between track and wire assuming wire perpendicular to z-axis
+  // calculate angle between track and wire assuming wire perpendicular to z-axis
   auto thetawire = asin((lexit.Y() - lentrance.Y()) / hitlengthZ);//check sign convention between O2 and Aliroot
 
-  // bfield
+  // local b-field
   double b[3] = {0., 0., 0.};
   double x[3] = {lentrance.X(), lentrance.Y(), lentrance.Z()};
   if (TGeoGlobalMagField::Instance()->GetField())  {
     TGeoGlobalMagField::Instance()->Field(x, b);
   } else{
     LOG(fatal) << "no b field in DEDigitizer";
-   // auto l3Current = ic.options().get<float>("l3Current");
-   // auto dipoleCurrent = ic.options().get<float>("dipoleCurrent");
-   // auto field =
-   // o2::field::MagneticField::createFieldMap(l3Current, dipoleCurrent, o2::field::MagneticField::kConvLHC, false, 3500.,
-   //                                          "A-A", "$(O2_ROOT)/share/Common/maps/mfchebKGI_sym.root");
-   //  field->Field(entrance, b);
   }
-  //calculate track betagamma
-  auto deltat = hit.exitTof() - hit.entranceTof();// delta t in nanoseconds
-  auto beta = std::sqrt((lexit.X() - lentrance.X()) * (lexit.X() - lentrance.X()) + (lexit.Y() - lentrance.Y()) * (lexit.Y() - lentrance.Y()) + hitlengthZ * hitlengthZ) / (deltat * 0.299792);
+  // calculate track betagamma, time in ns, space in cm
+  auto deltat = hit.exitTof() - hit.entranceTof();
+  auto beta = std::sqrt((lexit.X() - lentrance.X()) * (lexit.X() - lentrance.X()) + (lexit.Y() - lentrance.Y()) * (lexit.Y() - lentrance.Y()) + hitlengthZ * hitlengthZ) / (deltat * 29.9792);
 
   auto betagamma = beta * 1.0 / std::sqrt(1 - beta * beta);
   auto yAngleEffect = mResponse.inclandbfield(thetawire, betagamma, b[0]);

--- a/Detectors/MUON/MCH/Simulation/src/Response.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/Response.cxx
@@ -47,7 +47,6 @@ Response::Response(Station station)
   mChargeCorr = ResponseParam::Instance().chargeCorrelation;
   mChargeThreshold = ResponseParam::Instance().chargeThreshold;
 }
-
 //_____________________________________________________________________
 float Response::etocharge(float edepos) const
 {

--- a/Detectors/MUON/MCH/Simulation/src/Stepper.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/Stepper.cxx
@@ -59,6 +59,7 @@ void Stepper::process(const TVirtualMC& vmc)
   if (t.isEntering()) {
     float x, y, z;
     vmc.TrackPosition(x, y, z);
+    mTof = vmc.TrackTime();
     mEntrancePoint.SetXYZ(x, y, z);
     resetStep();
   }
@@ -68,9 +69,14 @@ void Stepper::process(const TVirtualMC& vmc)
 
   if (t.isExiting() || t.isStopped()) {
     float x, y, z;
+    double tof;
     vmc.TrackPosition(x, y, z);
+    tof = vmc.TrackTime();
+    //convert tof and mtof from s to ns before static cast as float
+    tof = 1e9 * tof;
+    mTof = 1e9 * tof; 
     mHits->emplace_back(stack->GetCurrentTrackNumber(), detElemId, mEntrancePoint,
-                        math_utils::Point3D<float>{x, y, z}, mTrackEloss, mTrackLength);
+                        math_utils::Point3D<float>{x, y, z}, mTrackEloss, mTrackLength, (float) mTof, (float) tof);
     resetStep();
   }
 }

--- a/Detectors/MUON/MCH/Simulation/src/Stepper.h
+++ b/Detectors/MUON/MCH/Simulation/src/Stepper.h
@@ -43,6 +43,7 @@ class Stepper
  private:
   float mTrackEloss{0.0};
   float mTrackLength{0.0};
+  double mTof{0.0};
   std::vector<o2::mch::Hit>* mHits{nullptr};
   math_utils::Point3D<float> mEntrancePoint;
 };

